openmc/deplete/stepresult.py

```diff
@@ -0,0 +1,526 @@
+"""The results module.
+
+Contains results generation and saving capabilities.
+"""
+
+from collections import OrderedDict
+import copy
+
+import h5py
+import numpy as np
+
+import openmc
+from openmc.mpi import comm, MPI
+from .reaction_rates import ReactionRates
+
+VERSION_RESULTS = (1, 1)
+
+
+__all__ = ["StepResult"]
+
+
+class StepResult:
+    """Result of a single depletion timestep
+
+    Attributes
+    ----------
+    k : list of (float, float)
+        Eigenvalue and uncertainty for each substep.
+    time : list of float
+        Time at beginning, end of step, in seconds.
+    source_rate : float
+        Source rate during timestep in [W] or [neutron/sec]
+    n_mat : int
+        Number of mats.
+    n_nuc : int
+        Number of nuclides.
+    rates : list of ReactionRates
+        The reaction rates for each substep.
+    volume : OrderedDict of str to float
+        Dictionary mapping mat id to volume.
+    mat_to_ind : OrderedDict of str to int
+        A dictionary mapping mat ID as string to index.
+    nuc_to_ind : OrderedDict of str to int
+        A dictionary mapping nuclide name as string to index.
+    mat_to_hdf5_ind : OrderedDict of str to int
+        A dictionary mapping mat ID as string to global index.
+    n_hdf5_mats : int
+        Number of materials in entire geometry.
+    n_stages : int
+        Number of stages in simulation.
+    data : numpy.ndarray
+        Atom quantity, stored by stage, mat, then by nuclide.
+    proc_time : int
+        Average time spent depleting a material across all
+        materials and processes
+
+    """
+    def __init__(self):
+        self.k = None
+        self.time = None
+        self.source_rate = None
+        self.rates = None
+        self.volume = None
+        self.proc_time = None
+
+        self.mat_to_ind = None
+        self.nuc_to_ind = None
+        self.mat_to_hdf5_ind = None
+
+        self.data = None
+
+    def __repr__(self):
+        t = self.time[0]
+        dt = self.time[1] - self.time[0]
+        return f"<StepResult: t={t}, dt={dt}, source={self.source_rate}>"
```

Comment:
Edge case if you create one of these bare and get an error before populating attributes: `time` and `source_rate` will be `None`

Might not be worth digging in to if this is more part of an internal API and likely not to be seen by users too much

Reply:
Yeah, this is mostly an internal thing and users should not be instantiating `StepResult` themselves. Fixing it would take a bit of restructuring that's beyond the scope of this PR

Reply:
Totally fair